kicker/main.c

```diff
@@ -316,195 +417,98 @@ ISR(SPI_STC_vect) {
-
-    // tx and led as output
-    // RX as intput
-    HAL_SetInputPin(BALL_SENSE_RX);
-    HAL_SetOutputPin(BALL_SENSE_TX);
-
-    // Enable the LED and TX until first loop
-    // This is because you cannot go from {input, tristate} -> {output, high}
-    // in a single step
-    HAL_SetPin(BALL_SENSE_TX);
-    
-    /**
-     * Enable SPI slave
-     */
-    // Assume default spi mode
-    // Allow interrupts, enable, and force slave SPI device
-    SPCR &= ~_BV(MSTR);
-    SPCR |= _BV(SPE) | _BV(SPIE);
-
-    ///////////////////////////////////////////////////////////////////////////
-    //  TIMER INITIALIZATION
-    //
-    //  The timer works by interrupt callback. The timer is based off of an
-    //  accumulator register that is incremented per clock tick. When the
-    //  accumulator register reaches the value in the target register,
-    //  the interrupt fires.
-    //
-    //  Initialization
-    //  1) The interrupt enable bit is set in TIMSK0
-    //  2) Clear the bit indicating timer matched the target/compare register
-    //  3) Set the value of the target/compare register
-    //
-    //  Callback
-    //  ISR(TIMER0_COMPA_vect)
-    //
-    //  Start/Global
-    //  kick()
-    //
-    //  initialize timer
-    TIMSK0 |= _BV(OCIE0A);    // Interrupt on TIMER 0
-    TCCR0A |= _BV(WGM01); // CTC
-    OCR0A = TIMING_CONSTANT;  // OCR0A is max val of timer before reset
-    ///////////////////////////////////////////////////////////////////////////
-
-
-    /**
-     * ADC Initialization
+    post_kick_cooldown_--;
+  } else if (kick_wait_ >= 0) {
+    /* POST KICK COOLDOWN
+     * enable charging
+     * don't allow kicking during the cooldown
      */
-    // Allow us to just read a single register to get the analog output
-    // instead of having to read 2 registers
-    // We don't care about the full 10 bit width, only the top 8
-    // Setup voltage monitor as input to adc
-    ADMUX |= _BV(ADLAR) | 0x00;//_BV(V_MONITOR_PIN.pin);
 
-    //  Ensure ADC isn't off
-    PRR &= ~_BV(PRADC);
+    // reenable charging
+    charge_allowed_ = true;
 
-    // Enable adc
-    ADCSRA |= _BV(ADEN);
-
-    /**
-     * Button logic
+    kick_wait_--;
+  } else {
+    /* IDLE/NOT RUNNING
+     * stop timer
      */
-    // latch debug state
-    in_debug_mode = !HAL_IsSet(DB_SWITCH);
 
-    // enable global interrupts
-    sei();
+    // stop prescaled timer
+    TCCR0 &= ~_BV(CS00);
+    // TCCR0 &= 0b00;
+  }
+}
+
+/*
+ * Executes a command that can come from SPI or a debug button
+ *
+ * WARNING: This will be called from an interrupt service routines, keep it
+ * short!
+ */
+uint8_t execute_cmd(uint8_t cmd) {
+  // if we don't change ret_val by setting it to voltage or
+  // something, then we'll just return the command we got as
+  // an acknowledgement.
+  bool allow_charge = !!(cmd & (1 << 4));
+  uint8_t kick_power = (cmd & 0xF) << 4;
+  uint8_t kick_activation = cmd & (3 << 5);
+  bool use_chip = !!(cmd & (1 << 7));
```

Comment:
Have the chip stuff implemented, but commented out. It'll make it easier later to come back and put it back in